### DISCUSSION TITLE
Hacking StumpWM with vim

### DISCUSSION THREAD

**Discussion by MaxGyver83** [stumpwm/stumpwm#1139] [Q&A] [answered]
I found this page in the StumpWM wiki: [Hacking StumpWM · stumpwm/stumpwm Wiki](https://github.com/stumpwm/stumpwm/wiki/Hacking-StumpWM)

Since I'm not an Emacs user, I tried start a Swank server as described:

```lisp
(ql:quickload :swank)
(swank:create-server)
```

and connect to it with vim + [slimv](https://github.com/kovisoft/slimv). But as far as I can tell, slimv seems not to see StumpWM's swank server and always starts a new one. I get a `CL-USER` prompt (instead of `STUMPWM-USER`).

Could anybody connect to StumpWM's Swank from vim?

**Comment by dmb2** [comment]
As long as stumpwm has a swank server that's listening (which I can confirm, I regularly connect via slime/sly) it should work.

I have had issues in the past with sly vs slime.  Sly uses slynk and slime uses swank.  Looking at the slimv page, it looks like there might be some special server-side sauce.  Can you try using this:
/.vim/bundle/slimv/slime/start-swank.lisp

and/or reproducing the logic to start swank within stumpwm (ie in your stumpwmrc).  Also, you might want to uninstall swank via quicklisp just to make sure they aren't interfering.

Finally... `(ql:quickload :swank)` in a sbcl repl *is not the same* as having it in your stumpwmrc.

**Comment by MaxGyver83** [comment]
> Finally... `(ql:quickload :swank)` in a sbcl repl _is not the same_ as having it in your stumpwmrc.

I entered the commands above after pressing `C-t :`. This should have the same effect as adding them to the `stumpwmrc`, right?

Anyway, next I tried adding those two lines to my `stumpwmrc` and I restarted: `C-t ; restart-hard` (I guess `C-t ; loadrc` is sufficient!?)

Then I can confirm that Swank seems to be running:

```shell
$ curl 127.0.0.1:4005
curl: (52) Empty reply from server
```

(I get `curl: (7) Failed to connect to 127.0.0.1 port 4005` when no swank server is running.)

I have activated logging in slimv. See the log file attached.

[swank.log](https://github.com/stumpwm/stumpwm/files/12515826/swank.log)

Can you see anything suspicious? If not, I'll create an issue for slimv.

**Comment by PuercoPop** [comment]
@MaxGyver83 What is the command that you use with slimv to connect to a server as opposed to start a server? I have tried locally the other vim extension vlime and have confirmed that with `<localleader>cc` one can connect vim to a swank server started by a running stumpwm.

**Reply by MaxGyver83** [reply]
I'm not sure about it. I think it happens here: [slimv/ftplugin/swank.py#L1277-L1283](https://github.com/kovisoft/slimv/blob/f4775f044cbb28ff4a444143cd2e6712b3694f21/ftplugin/swank.py#L1277-L1283)

**Reply by MaxGyver83** [reply]
Some debugging: 

```py
sock = socket.socket(socket.AF_INET, socket.SOCK_STREAM)
sock.connect(swank_server)
```

This causes a `socket.error`:

```
[Errno 111] Connection refused
```

**Reply by MaxGyver83** [reply]
Many more problems because SBCL was updated to version 2.3.8. After deleting the commonlisp cache, reinstalling swank, now starting StumpWM via roswell, I believe that now slimv connects to the StumpWM Swank server. I skip the part in slimv that starts a new swank server. And `pgrep -fa swank` doesn't return any match. But the REPL still starts in the CL-USER package:

```lisp
SBCL 2.3.7  Port: 4005  Pid: 19968
; SWANK 2.28
CL-USER> (setq *key-seq-color* "^3")
; in: SETQ *KEY-SEQ-COLOR*
;     (SETQ *KEY-SEQ-COLOR* "^3")
; 
; caught WARNING:
;   undefined variable: COMMON-LISP-USER::*KEY-SEQ-COLOR*
```

**Reply by PuercoPop** [reply]
Yeah, the REPL will likely start in in-cl_user, but you can change that with `(in-package #:stumpwm-user)`. The important thing is whether the REPL that you connected is the one that stump started. One quick way to check if by checking of the existence of the stumpwm package. ej. (find-package "STUMPWM")

**Reply by MaxGyver83** [reply]
OMG, the "stumpwm" package was found but I'm back at this issue: [Function sent to REPL with ,d is added to wrong package · Issue #132 · kovisoft/slimv](https://github.com/kovisoft/slimv/issues/132)

```lisp
CL-USER>(find-package "STUMPWM")
#<PACKAGE "STUMPWM">
CL-USER> (in-package #:stumpwm-user)
#<PACKAGE "STUMPWM-USER">
STUMPWM-USER> (setq *key-seq-color* "^4")
; in: SETQ *KEY-SEQ-COLOR*
;     (SETQ *KEY-SEQ-COLOR* "^4")
; 
; caught WARNING:
;   undefined variable: COMMON-LISP-USER::*KEY-SEQ-COLOR*
; 
; compilation unit finished
;   Undefined variable:
;     *KEY-SEQ-COLOR*
;   caught 1 WARNING condition
"^4"
STUMPWM-USER> 
```

**Reply by MaxGyver83** [reply]
But this time, deleting `~/.cache/common-lisp/sbcl-2.3.7-linux-x64/` doesn't fix it. I was guessing I had to reinstall swank for sbcl 2.3.7. The `sbcl` command is now at version 2.3.8. That's why I tried `Ctrl-t : (ql:quickload :swank)`. But slimv still gives me an `[Errno 111] Connection refused`. Doing the same in `ros run` (which is sbcl 2.3.7) doesn't help either. And `sbcl` (=2.3.8) is broken now:

```shell
$ sbcl --noinform
fatal error encountered in SBCL pid 13846 tid 13846:
core was built for runtime "4cb1d98433d6-root-2023-07-30-01-48-46" but this is "-builduser-2023-09-05-19-37-47"
```

When I run `stumpwm` (within StumpWM), I get this error:

```shell
$ stumpwm
Unhandled SIMPLE-ERROR in thread #<SB-THREAD:THREAD tid=14349 "main thread" RUNNING
                                    {10022C0003}>:
  Error opening shared object "/home/max/.cache/common-lisp/sbcl-2.3.7-linux-x64/home/max/quicklisp/dists/quicklisp/software/osicat-20230214-git/posix/libosicat.so":
  /home/max/.cache/common-lisp/sbcl-2.3.7-linux-x64/home/max/quicklisp/dists/quicklisp/software/osicat-20230214-git/posix/libosicat.so: cannot open shared object file: No such file or directory.
```

It looks like I'm lost in dependency hell :sob:

**Reply by MaxGyver83** [reply]
After quitting StumpWM and starting it again and adding `let g:slimv_lisp = 'ros run -- --no-userinit'` to `~/.vim/ftplugin/lisp.vim`, now I can connect to the correct Common Lisp image. Sending `(in-package :stumpwm-user)` updates the REPL prompt but obviously I'm still in the `:cl-user` package. When I enter `(in-package :stumpwm-user)` manually into the REPL and then `(setq *key-seq-color* "^4")`, the effect is indeed visible in StumpWM's help menu. So I think, it's only a slimv issue, maybe related to inconsistent SBCL and swank versions or a broken cache.

**Reply by dmb2** [reply]
Sounds like you figured it out? 

I love what roswell promises, but I never had it work well, I just remember getting stuck in subtle version mismatches that make it impossible to figure out who was loading what when.  Granted, I was much less experienced then so I'm not sure that was user ignorance or sharp edges on the tooling.

Once you get it figured out start stripping back to basics.  I eventually settled on just sbcl installed via the OS package manager, quicklisp for external packages and git for stumpwm only.  I tried sly for a while and again and ran into swank issues (it uses slynk) so there is some inherent challenges to making versions work together.

**Reply by MaxGyver83** [reply]
Thanks. Good to hear I'm not the only one struggling with this. I definitely can't leave it like it is because sending expressions from vim to the REPL now doesn't work (only typing them directly into the REPL).

I also would like to do it like you and only use Arch's SBCL.

Some more (Common Lisp) beginner questions:

1. AFAIU, my `stumpwm` binary is a full Common Lisp image/core. Does this mean it contains a SBCL 2.3.7?
2. Is it necessary then that I install SBCL 2.3.7 on my system? I guess, I don't need an SBCL installation at all, right?
3. Is quicklisp connected to an SBCL version?
4. And the packages I install via quicklisp?
5. And does StumpWM read my `~/.sbclrc`?

**Comment by MaxGyver83** [comment, marked as answer]
In case anybody else finds this, these were my problems:

* sbcl 2.3.8 is somehow broken on my system. A re-installation doesn't fix it. Now I'm using sbcl 2.3.7 installed via _roswell_. My _slimv_ settings now:
```vim
let g:slimv_swank_cmd = "!ros -e '(ql:quickload :swank) (swank:create-server)' wait &"
let g:slimv_lisp = 'ros run'
let g:slimv_impl = 'sbcl'
```
* The SBCL cache was broken. This can be fixed by deleting it with p.e. `rm -rf ~/.cache/common-lisp/sbcl-2.3.7-linux-x64/`.
* I had created `~/.vim/ftplugin/lisp.vim` for my slimv and paredit settings. This caused that vim system's lisp plugin wasn't loaded and _slimv_ didn't work correctly. Fixed by moving `~/.vim/ftplugin/lisp.vim` to `~/.vim/after/ftplugin/lisp.vim`.
* The StumpWM wiki recommends (Emacs users) to add `(swank:create-server)` to the _stumpwmrc_. With this setting, I can connect to StumpWM's swank server only once per StumpWM session. This means, after closing vim and later opening it again, I can't connect anymore. With `(swank:create-server :dont-close t)`, I can disconnect and reconnect multiple times.
